azure_scripts/lotka_volterra.py

```diff
@@ -0,0 +1,269 @@
+parser.add_argument("--model_seed", type=int, default=0)
+args = parser.parse_args()
+
+if ON_AZURE:
+  run.log('batch_size', args.batch_size)
+  run.log('n_simulations', args.n_simulations)
+  run.log('n_epochs', args.n_epochs)
+  run.log('dimension', args.dimension)
+  run.log('bijector_layers_size', args.bijector_layers_size)
+  run.log('bijector_layers_shape', args.bijector_layers_shape)
+  run.log('nf_layers', args.nf_layers)
+  run.log('n_components', args.n_components)
+  run.log('score_weight', args.score_weight)
+  run.log('model_seed', args.model_seed)
+else:
+  print(args)
+
+# create truth
+key = jax.random.PRNGKey(0)
+lokta_volterra_10 = partial(lotka_volterra, ts=jnp.linspace(0, 18.9, 5))
+lv_cond = condition(lokta_volterra_10, {'z': jnp.array([30.0, 1.0])})
+(_, samples_0), _ = get_samples_and_scores(lv_cond, key, batch_size=1)
+observation = np.reshape(samples_0['y'], (-1, 10), order='F')
+truth_0 = samples_0['theta']
+
+# create data train
+key = jax.random.PRNGKey(10)
+(_, samples), score = get_samples_and_scores(lv_cond, key, batch_size=args.n_simulations)
+batch = np.reshape(samples['y'], (-1, 10), order='F')
+mu = samples['theta']
+
+# normalize data
+scale_theta = jnp.std(mu, axis=0) / 0.02
+shift_theta = jnp.mean(mu/scale_theta, axis=0) - 0.4
+transformation_params = tfb.Chain(
+    [tfb.Scale(scale_theta), tfb.Shift(shift_theta)])
+
+scale_reg = jnp.std(batch, axis=0) / 0.01
+shift_reg = jnp.mean(batch/scale_reg, axis=0) - 0.4
+transformation_x = tfb.Chain(
+  [tfb.Scale(scale_reg), tfb.Shift(shift_reg)])
+
+normalized_p = transformation_params.inverse(mu)
+normalized_reg = transformation_x.inverse(batch)
+
+# create model
+bijector_layers = [args.bijector_layers_size] * args.bijector_layers_shape
+
+bijector = partial(AffineSigmoidCoupling, layers=bijector_layers, n_components=args.n_components, activation=jax.nn.silu)
+NF = partial(ConditionalRealNVP, n_layers=args.nf_layers, bijector_fn=bijector)
+
+nvp_nd = hk.without_apply_rng(hk.transform(lambda p, x: NF(args.dimension)(x).log_prob(p).squeeze()))
+nvp_sample_nd = hk.transform(lambda x: NF(args.dimension)(x).sample(10000, seed=hk.next_rng_key()))
+
+# init parameters
+rng_seq = hk.PRNGSequence(args.model_seed)
+params_nd = nvp_nd.init(next(rng_seq), 0.4 * jnp.ones([1, 4]), 0.4 * jnp.ones([1, 10]))
+
+# init optimizer
+scheduler = optax.exponential_decay(
+    init_value=0.001, transition_steps=2000, decay_rate=0.9, end_value=0.00001)
+optimizer = optax.chain(
+  optax.scale_by_adam(), optax.scale_by_schedule(scheduler), optax.scale(-1))
+opt_state = optimizer.init(params_nd)
+
+# define loss function and model update
+def loss_fn(params, weight, mu, batch, score):
+  log_prob, out = jax.vmap(
+    jax.value_and_grad(
+      lambda theta, x: nvp_nd.apply(params, theta.reshape([1, 4]), x.reshape([1, 10])).squeeze()
+    )
+  )(mu, batch)
+  return -jnp.mean(log_prob) + weight * jnp.mean(jnp.sum((out - score)**2, axis=1))
```

Comment:
NaNs come from there if you comment `+ weight * jnp.mean(jnp.sum((out - score)**2, axis=1))` it works. For some reason, the NF gradient looks like this at some point <img width="940" alt="image" src="https://user-images.githubusercontent.com/72011736/167869890-77a1e554-6507-49fe-8eb7-a489a528d0fa.png">